### DISCUSSION TITLE
Linking directly to a page or a content chunk

### DISCUSSION THREAD

**Discussion by iturgeon** [ucfopen/Obojobo#1750] [Ideas]
Faculty would like to be able to link directly to a section in a module.  We should explore the idea.

A faculty member's question:

> if we want to link to a discussion module from within the Obojobo module (to have them to out and complete an activity, then come back to Obojobo), can we get a bookmark back to the specific part of the Obojobo module? For now it opens the discussion post (or other Canvas module) in a separate browser tab, and the student can complete it there, which is fine, but to navigate back to Obojob they have to find the tab they had the Obojobo module open in and manually navigate back to it. I tried just hyperlinking it but it took me back to the beginning of the Obojobo module. Sometimes I got an error message saying I'd already completed that part of the module, as well.

* what are the implications with LTI launches, resuming, etc?
* what happens if the student doesn't have an active session and clicks the link?
* how would faculty build the url needed?

**Comment by zachberry** [comment]
NavStore has some variables that might be useful - itemsByPath and itemsByFullPath, which were intended to be some of the plumbing to get this to work. However that's fairly old now, I don't want to force us to go down that road if there's a simpler solution

**Comment by zachberry** [comment]
There's a few different problems to solve (perhaps others too)

1. Being able to refresh the page of an obo module and return to the page/location you were at (currently you are kicked back to the start page, despite the URL seeming to promise otherwise)
2. External link to a specific page in an obo module
3. Internal link inside a module to link to a specific page in that module
4. Link to a specific node inside a module
5. Link to a selection of text inside a node inside a module
6. Module to module link - Module A links to a page in module B


* NavStore already updates the URL with history.pushState. 1 should be solvable by history.popState, but need to make sure they aren't navigating to somewhere forbidden (aka if navigation is locked, in an assessment, etc)
* Internal linking would automatically solved once 1 is solved, however it'd be nicer to do some magic. We could know when an author puts in an internal link and shorten it. We could also create an editor UI to let them link to a page. We might also need a system that would update links if the URL of a page changes.
* External link to a module would also be solved with the solutions above
* Linking to a node might be of the form `#node-id`. Focus Store has methods to focus on a node, so once a page is loaded focus store could be told to focus on the node indicated by the given node id. We may need to add something in the (...) menu to be able to copy a link to a node (or, do something in the Create/Edit link UI)
* Linking to a selection of text is also possible, using the data structures of the Viewer. Perhaps a URL of `#start=node-id,text-group-index,start&end=node-id,text-group-index,end` (so for example `#start=my-heading,0,4&end=my-text,1,5` which would select starting at the node with id `my-heading` at the 4th character to the node with id `my-text`, on the second line, fifth character. (I'm not sold on this syntax, the gist is we could somehow describe a selection range in the URL)
* Module to module link may not be possible without LTI work

**Comment by zachberry** [comment]
Some inspiration for inter-document linking - here's Google Doc's newer (new to me anyway) link menu:

<img width="535" alt="Screen Shot 2021-03-26 at 9 55 25 AM" src="https://user-images.githubusercontent.com/73479/112642298-89134480-8e19-11eb-943e-2e72fe7265e6.png">

<img width="564" alt="Screen Shot 2021-03-26 at 9 55 20 AM" src="https://user-images.githubusercontent.com/73479/112642313-8dd7f880-8e19-11eb-83b2-08ac691f4987.png">
